src/text/raw_text_reader.rs

```diff
@@ -249,25 +266,52 @@ impl<T: TextIonDataSource> RawTextReader<T> {
         }
     }
 
+    fn parse_next<P, O>(&mut self, parser: P) -> IonResult<Option<O>>
+    where
+        P: Fn(&str) -> IonParseResult<O>,
+    {
+        match self.parse_next_nom(parser) {
+            RootParseResult::Ok(item) => Ok(Some(item)),
+            RootParseResult::Eof => Ok(None),
+            RootParseResult::NoMatch => {
+                // Return an error that contains the text currently in the buffer (i.e. what we
+                // were attempting to parse.)
+                let error_message = format!(
+                    "unrecognized input near line {}: '{}'",
+                    self.buffer.lines_loaded(),
+                    self.buffer.remaining_text(),
+                );
+                decoding_error(error_message)
+            }
+            RootParseResult::Failure(error_message) => decoding_error(error_message),
+        }
+    }
+
     /// Attempts to parse the next entity from the stream using the provided parser.
     /// If there isn't enough data in the buffer for the parser to match its input conclusively,
     /// more data will be loaded into the buffer and the parser will be called again.
     /// If EOF is encountered, returns `Ok(None)`.
-    fn parse_next<P, O>(&mut self, parser: P) -> IonResult<Option<O>>
+    fn parse_next_nom<P, O>(&mut self, parser: P) -> RootParseResult<O>
```

Comment:
🗺️ Unlike `parse_next`, `parse_next_nom` distinguishes between fatal and non-fatal errors.